[FX] Fix create_arg for NamedTuple

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48986 [FX] Fix create_arg for NamedTuple**

Closes https://github.com/pytorch/pytorch/issues/48813

Previously, constructing a NamedTuple with `type(a)(self.create_arg(elem) for elem in a)` would fail, saying `__new__() missing 1 required positional argument: 'aux_logits'`, ostensibly because (unlike `tuple()`) a `NamedTuple`'s constructor does not accept a generator expression as input. This adds a special case for `NamedTuple`s in `create_arg`

Differential Revision: [D25387156](https://our.internmc.facebook.com/intern/diff/D25387156)

Commit:
[NT] Fix create_arg for NamedTuple

[ghstack-poisoned]

Commit:
Update on "[NT] Fix create_arg for NamedTuple"


Closes https://github.com/pytorch/pytorch/issues/48813

Previously, constructing a NamedTuple with `type(a)(self.create_arg(elem) for elem in a)` would fail, saying `__new__() missing 1 required positional argument: 'aux_logits'`, ostensibly because (unlike `tuple()`) a `NamedTuple`'s constructor does not accept a generator expression as input. This adds a special case for `NamedTuple`s in `create_arg`

Differential Revision: [D25387156](https://our.internmc.facebook.com/intern/diff/D25387156)

[ghstack-poisoned]

zdevito commented on torch/fx/proxy.py:
```diff
@@ -51,7 +51,14 @@ def create_arg(self, a: Any) -> Argument:
         # aggregates
         if isinstance(a, (tuple, list)):
-            return type(a)(self.create_arg(elem) for elem in a)
+            if hasattr(a, '_fields'):
```
Seems fine, but I'd split the name tuple isinstance check to the top level to make it more clear.
`if isinstance(x, named_tuple):`

zdevito commented on torch/fx/proxy.py:
```diff
@@ -51,7 +51,14 @@ def create_arg(self, a: Any) -> Argument:
+                # expression as an argument to their constructor, so build this
+                # intermediate tuple and unpack it into the NamedTuple constructor
+                args = tuple(self.create_arg(elem) for elem in a)
+                return type(a)(*args)
```
`create_arg` should be narrowing the set of objects allowed to those supported by `map_arg`, but NamedTuple won't be supported by map arg. Is the intension to add named tuples to map_arg? If so that should also be added, if not, then this should just erase the naming from the tuple.

Commit:
Update on "[FX] Fix create_arg for NamedTuple"


Closes https://github.com/pytorch/pytorch/issues/48813

Previously, constructing a NamedTuple with `type(a)(self.create_arg(elem) for elem in a)` would fail, saying `__new__() missing 1 required positional argument: 'aux_logits'`, ostensibly because (unlike `tuple()`) a `NamedTuple`'s constructor does not accept a generator expression as input. This adds a special case for `NamedTuple`s in `create_arg`

Differential Revision: [D25387156](https://our.internmc.facebook.com/intern/diff/D25387156)

[ghstack-poisoned]
